servicetalk-concurrent-api/docs/modules/ROOT/pages/blocking-implementation.adoc

```diff
@@ -63,34 +103,21 @@ Taking the same example from xref:{page-version}@servicetalk-concurrent-api::blo
        .map(resp -> {
             return resp.toString(); # <2>
        })
-       .flatMap(stringResp -> { # <3>
+       .publishOn(executor) # <3>
+       .flatMap(stringResp -> { # <4>
             return client2.request(stringResp);
        })
        .filter(stringResp -> {
-            stringResp.equals("Hello World");  # <4>
+            stringResp.equals("Hello World");  # <5>
        });
 ----
 <1> A hypothetical client which provides a `request()` method that returns a `Single<Response>`.
 <2> Converting the response to a `String`.
-<3> Call another `client2` that provides a new `Single` which is returned from `flatMap`.
-<4> Only allow "Hello World" messages to be emitted.
-
-In the above example the operators `map` and `filter` will not wrap `Subscriber` and `Subscription` since they do not do
-any asynchronous work. However, `flatmap` will wrap `Subscriber` and `Subscription` to offload the calls to the chosen
-`Executor`.
-
-== Why should we wrap `Subscriber` and `Subscription`?
-
-There are two places we would wrap `Subscriber` and `Subscription`:
-
-1. Original asynchronous sources.
-2. Asynchronous operators. eg: `flatMap`
-
-Since every asynchronous source is associated with an `Executor`, it is required to use the `Executor` for interacting
-with `Subscriber` and `Subscription`.
-
-If we do not wrap for asynchronous operators then in the above example, operator `filter` (4) will be invoked in the
-`Executor` defined by `client2` inside the `flatmap`. This may lead to inadvertent and not initially obvious
-blocking of an event loop thread. Consider a scenario where `client2` executes user code on an event loop thread, but
-the original `client` executes user code on an application thread (which allows blocking). In this scenario it may look
-like blocking code is OK at point (4) above, but that will actually result in blocking `client2`'s event loop thread.
+<3> Offload execution to the provided `Executor`
+<4> Call another `client2` that provides a new `Single` which is returned from `flatMap`.
+<5> Only allow "Hello World" messages to be emitted.
+
+In the above example the operators `map` and `filter` do not need to offload since they do not do
+any asynchronous work. However, `flatmap` should be offloaded since it asynchronously executes another request. Adding
```

Comment:
`However, flatmap should be offloaded since it asynchronously executes another request.` -> this is just example code which decides where offloading is used, but this statement implies that offloading is required bcz another request is made (which isn't necessary). I think this is a hold over from previous examples where executor affinity would offload the resulting single, but we no longer do this.

Consider clarifying the control flow like is done for the below example (feel free to use w/e example):

```java
   Publisher.range(1, 500)
             .map(Element::new)
             .publishOn(smallExecutor)
             .map(myItem -> {
                 sleep(10);
                 System.out.println("\nOn Thread " + Thread.currentThread().getName() + " Received MyItem " + myItem.getValue());
                 return myItem;
             })
             .subscribeOn(biggestExecutor)
             .toFuture().get();
```


> 1. toFuture().get() will do a subscribe(Subscriber), which flows up the operator chain (subscribeOn (offload onto thread from biggestExecutor) -> map -> publishOn -> map -> Range).
> 2. Range will call Subscriber.onSubscribe(Subscription) on the Subscriber, which flows back down the operator chain (Range -> map -> publishOn (offloads to thread from smallestExecutor) -> map -> subscribeOn -> toFuture).
> 3. toFuture will call Subscription.request(Long.MAX_VALUE), which flows up the operator chain (subscribeOn (offloads onto thread from smallestExecutor -> map -> publishOn -> map -> Range).
> 4. Range will do onNext for 1-500 items synchronously (on a thread from smallestExecutor), where each onNext flows back down the operator chain (Range -> map -> publishOn (offloads to thread from smallestExecutor) -> map -> subscribeOn -> toFuture).

Reply:
The diagnostics `whenRequest` are nice for folks to understand how to debug/diagnose but isn't necessary for someone first seeing the code, and can be intimidating/off-putting. I would suggest two things:
- deliver the first code with simple text explanations to keep code content minimal/focused.
- having the combined publishOn/subscribeOn example is still useful so folks can see how these operators interact, but also consider delivering 2 simple code examples with each independently if this makes the concepts easier to explain (and lightens the combined explanation)
- have a followup "see for your self with instrumentation" section that includes the `when*` printouts as described above.